File: trees/101_symmetric_tree.py
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def is_symmetric_stack(root: TreeNode) -> bool:
    stack = [(root.left, root.right)]
    while stack:
        left, right = stack.pop()
        if not left and not right:
            continue
        if not left or not right or left.val != right.val:
            return False

        stack.append((left.left, right.right))
        stack.append((left.right, right.left))

    return True


def is_symmetric_queue(root: TreeNode) -> bool:
    queue = deque([(root.left, root.right)])
    while queue:
        left, right = queue.popleft()
        if not left and not right:
            continue
        if not left or not right or left.val != right.val:
            return False

        queue.append((left.left, right.right))
        queue.append((left.right, right.left))

    return True

def is_symmetric_recursive(self, root: TreeNode) -> bool:
    def dfs(left, right):
        if not left and not right:
            return True

        if not left or not right or left.val != right.val:
            return False

        return dfs(left.left, right.right) and dfs(left.right, right.left)

    return dfs(root.left, root.right)

File: trees/101_symmetric_tree.go
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type Stack[T any] []T

// удаляет и возвращает последний элемент стека
func (s *Stack[T]) Pop() T {
	pop := (*s)[len(*s)-1]
	*s = (*s)[:len(*s)-1]
	return pop
}

// возвращает последний элемент стека
func (s Stack[T]) Top() T {
	return s[len(s)-1]
}

// добавляет элемент в стэк
func (s *Stack[T]) Append(val T) {
	*s = append(*s, val)
}

type Pair struct {
	Left  *TreeNode
	Right *TreeNode
}

// ИТЕРАТИВНО (также через очередь)
func isSymmetric(root *TreeNode) bool {
	stack := Stack[Pair]{Pair{root.Left, root.Right}}

	for len(stack) > 0 {
		pair := stack.Pop()

		if pair.Left == nil && pair.Right == nil {
			continue
		}
		if pair.Left == nil || pair.Right == nil {
			return false
		}

		if pair.Left.Val != pair.Right.Val {
			return false
		}

		stack.Append(Pair{pair.Left.Right, pair.Right.Left})
		stack.Append(Pair{pair.Left.Left, pair.Right.Right})
	}

	return true
}

// РЕКУРСИВНО
func dfs(left, right *TreeNode) bool {
	if left == nil && right == nil {
		return true
	}

	if (left == nil || right == nil) || (left.Val != right.Val) {
		return false
	}

	return dfs(left.Left, right.Right) && dfs(left.Right, right.Left)
}

func isSymmetric(root *TreeNode) bool {
	return dfs(root.Left, root.Right)
}